File: lib/ai/rag.ts
import { generateCompletion, streamCompletion } from './providers'
import { ContextEngine } from './context-engine'
import type { KnowledgeItem } from '@/types'

interface RAGResponse {
    answer: string
    sources: KnowledgeItem[]
}

/**
 * RAG (Retrieval Augmented Generation) Layer
 * Combines semantic search with LLM generation for contextual answers
 */
export class RAG {
    private static readonly SYSTEM_PROMPT = `You are the user's Second Brain assistant. You help them recall and connect their knowledge.

RULES:
1. Answer ONLY using the provided notes as context
2. If the notes don't contain relevant information, say "I don't have notes about that"
3. Quote note titles as sources when referencing information
4. Be concise but thorough
5. Make connections between related notes when relevant`

    /**
     * Generate answer using retrieved context
     */
    static async query(
        question: string,
        userId: string,
        activeNote?: KnowledgeItem
    ): Promise<RAGResponse> {
        // Get relevant context
        const context = await ContextEngine.getContext(question, userId, activeNote)

        if (context.notes.length === 0) {
            return {
                answer: "I don't have any notes that could help answer this question. Try capturing some knowledge first!",
                sources: [],
            }
        }

        // Build context string
        const contextStr = this.buildContextString(context.notes)

        // Generate answer
        const prompt = `${this.SYSTEM_PROMPT}

CONTEXT (User's Notes):
${contextStr}

USER QUESTION: ${question}

Answer the question using ONLY the context above. Cite sources by title.`

        const answer = await generateCompletion(prompt, {
            provider: 'groq',
            temperature: 0.5,
            maxTokens: 1024,
        })

        return {
            answer,
            sources: context.notes,
        }
    }

    /**
     * Stream answer for real-time responses
     */
    static async *streamQuery(
        question: string,
        userId: string,
        activeNote?: KnowledgeItem
    ): AsyncGenerator<{ chunk?: string; sources?: KnowledgeItem[] }, void, unknown> {
        // Get relevant context
        const context = await ContextEngine.getContext(question, userId, activeNote)

        if (context.notes.length === 0) {
            yield {
                chunk: "I don't have any notes that could help answer this question.",
                sources: [],
            }
            return
        }

        // Yield sources first
        yield { sources: context.notes }

        // Build context string
        const contextStr = this.buildContextString(context.notes)

        // Stream answer
        const prompt = `${this.SYSTEM_PROMPT}

CONTEXT (User's Notes):
${contextStr}

USER QUESTION: ${question}

Answer the question using ONLY the context above. Cite sources by title.`

        for await (const chunk of streamCompletion(prompt, {
            provider: 'groq',
            temperature: 0.5,
            maxTokens: 1024,
        })) {
            yield { chunk }
        }
    }

    /**
     * Build context string from notes
     */
    private static buildContextString(notes: KnowledgeItem[]): string {
        return notes
            .map((note, i) => {
                const summary = note.summary || note.content.slice(0, 500)
                const tags = note.tags?.length ? `[Tags: ${note.tags.join(', ')}]` : ''
                return `--- Note ${i + 1}: "${note.title}" ${tags} ---
${summary}
`
            })
            .join('\n')
    }

    /**
     * Summarize a single note
     */
    static async summarize(content: string): Promise<string> {
        const prompt = `Summarize the following content in 2-3 concise sentences. Focus on key insights and actionable points.

CONTENT:
${content}

SUMMARY:`

        return generateCompletion(prompt, {
            provider: 'groq',
            temperature: 0.3,
            maxTokens: 256,
        })
    }

    /**
     * Generate tags for content
     */
    static async generateTags(title: string, content: string): Promise<string[]> {
        const prompt = `Generate 3-5 relevant tags for this knowledge item. Return ONLY a JSON array of lowercase tags.

TITLE: ${title}

CONTENT: ${content.slice(0, 1000)}

TAGS (JSON array only, no other text):`

        const response = await generateCompletion(prompt, {
            provider: 'groq',
            temperature: 0.3,
            maxTokens: 100,
        })

        try {
            // Extract JSON array from response
            const match = response.match(/\[.*\]/)
            if (match) {
                return JSON.parse(match[0])
            }
        } catch {
            console.error('Failed to parse tags:', response)
        }

        return []
    }
}

File: backend/app/services/rag.py
"""
RAG Engine — Retrieval Augmented Generation with User Context.

📚 LEARNING:
    RAG = Retrieve relevant context + Generate answer using that context.

    Flow:
    1. User asks: "What did I learn about React?"
    2. RETRIEVE: Search vector DB with the question text → get top 5 matching notes
    3. AUGMENT: Format those notes + user context as context for the LLM
    4. GENERATE: LLM answers using ONLY the retrieved context

    User Context:
    The user's profile (interests, role, goals) is injected into the system
    prompt so the AI gives personalized, tailored answers.
"""

import json
import re
from app.services.llm import llm_service
from app.db.vector_store import vector_store
from app.db.firebase import firebase_db


BASE_SYSTEM_PROMPT = """You are the user's Second Brain assistant called Axon.
You help them recall and connect their knowledge.

RULES:
1. Answer ONLY using the provided notes as context
2. If the notes don't contain relevant information, say "I don't have notes about that"
3. Quote note titles as sources when referencing information
4. Be concise but thorough
5. Make connections between related notes when relevant"""


class RAGEngine:
    """Core RAG pipeline: search → retrieve → generate (with user context)."""

    def _build_system_prompt(self, user_context_str: str = "") -> str:
        """Build system prompt with optional user context."""
        prompt = BASE_SYSTEM_PROMPT
        if user_context_str:
            prompt += f"\n\nUSER PROFILE: {user_context_str}\nTailor your response style and focus areas based on this profile."
        return prompt

    async def query(self, question: str, user_id: str) -> dict:
        """Full RAG pipeline with user context injection."""

        # Step 0: Fetch user context for personalized responses
        user_context = await firebase_db.get_user_context(user_id)
        user_context_str = user_context.to_prompt()

        # Step 1: Search vector store with question text
        results = vector_store.search(question, user_id, n_results=5)

        if not results:
            return {
                "answer": "I don't have any notes that could help answer this. Try adding some knowledge items first!",
                "sources": [],
            }

        # Step 2: Fetch full note content from Firebase
        sources = []
        context_parts = []
        for i, result in enumerate(results):
            note = await firebase_db.get_knowledge_item_by_id(result["id"])
            if note:
                sources.append({
                    "id": note["id"],
                    "title": note["title"],
                    "type": note["type"],
                    "summary": note.get("summary"),
                })
                summary = note.get("summary") or note["content"][:500]
                tags = note.get("tags", [])
                tag_str = f" [Tags: {', '.join(tags)}]" if tags else ""
                context_parts.append(
                    f'--- Note {i + 1}: "{note["title"]}"{tag_str} ---\n{summary}'
                )

        context_str = "\n\n".join(context_parts)

        # Step 3: Generate answer with user-personalized system prompt
        system_prompt = self._build_system_prompt(user_context_str)

        prompt = f"""CONTEXT (User's Notes):
{context_str}

USER QUESTION: {question}

Answer using ONLY the context above. Cite sources by title."""

        answer = await llm_service.generate(
            prompt=prompt,
            system_prompt=system_prompt,
            temperature=0.5,
            max_tokens=1024,
        )

        return {"answer": answer, "sources": sources}

    async def summarize(self, content: str) -> str:
        """Generate a concise summary of content."""
        prompt = (
            "Summarize the following content in 2-3 concise sentences. "
            "Focus on key insights and actionable points.\n\n"
            f"CONTENT:\n{content}\n\nSUMMARY:"
        )
        return await llm_service.generate(prompt, temperature=0.3, max_tokens=256)

    async def auto_tag(self, title: str, content: str) -> list[str]:
        """Generate relevant tags for content."""
        prompt = (
            "Generate 3-5 relevant tags for this knowledge item. "
            "Return ONLY a JSON array of lowercase tags.\n\n"
            f"TITLE: {title}\n"
            f"CONTENT: {content[:1000]}\n\n"
            "TAGS (JSON array only):"
        )
        response = await llm_service.generate(prompt, temperature=0.3, max_tokens=100)

        try:
            match = re.search(r"\[.*\]", response)
            if match:
                return json.loads(match.group())
        except Exception:
            print(f"Failed to parse tags: {response}")

        return []


# ── Singleton ──────────────────────────────────────────────────
rag_engine = RAGEngine()
